src/meta/meta_ckpt.c

```diff
@@ -1347,18 +1515,255 @@ __wt_meta_sysinfo_set(WT_SESSION_IMPL *session)
       txn->snap_min, txn->snap_max, txn->snapshot_count,
       __wt_timestamp_to_string(txn_global->oldest_timestamp, ts_string[0]),
       __wt_timestamp_to_string(txn_global->meta_ckpt_timestamp, ts_string[1]),
-      S2C(session)->base_write_gen);
+      conn->base_write_gen);
+
+    /* Record the base write gen in metadata as part of full checkpoints. */
 
-    /* Record the base write gen in metadata as part of checkpoint */
-    WT_ERR(__wt_buf_fmt(
-      session, buf, WT_SYSTEM_BASE_WRITE_GEN "=%" PRIu64, S2C(session)->base_write_gen));
-    WT_ERR(__wt_metadata_update(session, WT_SYSTEM_BASE_WRITE_GEN_URI, buf->data));
+    if (full) {
+        WT_ERR(__wt_buf_fmt(
+          session, valbuf, WT_SYSTEM_BASE_WRITE_GEN "=%" PRIu64, conn->base_write_gen));
+        WT_ERR(__wt_metadata_update(session, WT_SYSTEM_BASE_WRITE_GEN_URI, valbuf->data));
+    }
 
 err:
-    __wt_scr_free(session, &buf);
+    __wt_scr_free(session, &valbuf);
+    if (name != NULL)
+        __wt_scr_free(session, &uribuf);
+    return (ret);
+}
+
+/*
+ * __wt_meta_sysinfo_clear --
+ *     Clear the system information (for a named checkpoint) from the metadata.
+ */
+int
+__wt_meta_sysinfo_clear(WT_SESSION_IMPL *session, const char *name, size_t namelen)
+{
+    WT_DECL_ITEM(uribuf);
+    WT_DECL_RET;
+
+    WT_RET(__wt_scr_alloc(session, namelen + 128, &uribuf));
+
+    WT_ERR(__meta_sysinfo_remove(session, false, name, namelen, uribuf, WT_SYSTEM_CKPT_URI));
+    WT_ERR(__meta_sysinfo_remove(session, false, name, namelen, uribuf, WT_SYSTEM_OLDEST_URI));
+    WT_ERR(
+      __meta_sysinfo_remove(session, false, name, namelen, uribuf, WT_SYSTEM_CKPT_SNAPSHOT_URI));
+
+err:
+    __wt_scr_free(session, &uribuf);
     return (ret);
 }
 
+/*
+ * __wt_meta_read_checkpoint_snapshot --
+ *     Fetch the snapshot data for a checkpoint from the metadata file. Reads the selected named
+ *     checkpoint's snapshot, or if the checkpoint name passed is null, the most recent checkpoint's
+ *     snapshot if the checkpoint name passed is null. The snapshot list returned is allocated and
+ *     must be freed by the caller.
+ */
+int
+__wt_meta_read_checkpoint_snapshot(WT_SESSION_IMPL *session, const char *ckpt_name,
+  uint64_t *snap_min, uint64_t *snap_max, uint64_t **snapshot, uint32_t *snapshot_count,
+  uint64_t *ckpttime)
+{
+    WT_CONFIG list;
+    WT_CONFIG_ITEM cval;
+    WT_CONFIG_ITEM k;
+    WT_CONNECTION_IMPL *conn;
+    WT_DECL_ITEM(tmp);
+    WT_DECL_RET;
+    uint32_t counter;
+    char *sys_config;
+
+    conn = S2C(session);
+    counter = 0;
+    sys_config = NULL;
+
+    /*
+     * There's an issue with checkpoints produced by some old versions having bad snapshot data.
+     * (See WT-8395.) We should ignore those snapshots when we can identify them. This only applies
```

Comment:
This one's already mentioned in txn_checkpoint.c; that's where I got it from (doubt I would have managed to find it otherwise...) so we think it's probably ok.

Reply:
Agreed.  And there are plenty of other places in WT code that do this, so it doesn't seem to be a hard and fast rule. So I don't have a good sense of how strongly the team feels about it.  Hence I mentioned it to see if anyone else jumped in to say "yeah, don't do that."  But all quiet, so I'm good with this.